crates/driver/src/domain/time.rs

```diff
@@ -0,0 +1,60 @@
+use {
+    crate::infra::{
+        solver::Timeouts,
+        {self},
+    },
+    thiserror::Error,
+};
+
+/// Deadlines for different parts of the driver execution.
+/// The driver is expected to return the solution to the autopilot before the
+/// driver deadline.
+/// The solvers are expected to return the solution to the driver before the
+/// solvers deadline.
+#[derive(Copy, Clone, Debug, Default)]
+pub struct Deadline {
+    driver: chrono::DateTime<chrono::Utc>,
+    solvers: chrono::DateTime<chrono::Utc>,
+}
+
+impl Deadline {
+    pub fn new(deadline: chrono::DateTime<chrono::Utc>, timeouts: Timeouts) -> Self {
+        let deadline = deadline - timeouts.http_delay;
+        Self {
+            driver: deadline,
+            solvers: {
+                let now = infra::time::now();
+                let duration = deadline - now;
+                now + duration * (timeouts.solving_share_of_deadline.get() * 100.0).round() as i32
+                    / 100
+            },
+        }
+    }
+
+    /// Remaining time until the deadline for driver to return solution to
+    /// autopilot is reached.
+    pub fn remaining_for_driver(self) -> Result<chrono::Duration, DeadlineExceeded> {
```

Comment:
nit: I personally like longer names but given in the rest of the codebase we try no longer to use compound function names where possible, should these become just driver and solver (and have the fact that this is the remaining duration be conveyed by the return type)?

Reply:
Makes sense. Fixed.